### PR TITLE
Custom distribution and its analysis

### PR DESCRIPTION
**Pull request by Prateek-Ch**
Aim was to be able to input a custom relation distribution and based on that have the dataset sampled, model trained and then use that model on test set.

Some important results

1. Original Dataset Splits

2. Custom Dataset Splits

3. Custom Dataset Results

### CODE REVIEW

File: kge/util/distribution-shift/custom_distribution.py [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+    def __init__(self, checkpoint_path, noise_magnitude):
+        _, self.dataset = dsutils.load_model_and_dataset(checkpoint_path)
+        self.noise_magnitude = noise_magnitude
+
+    def calculate_target_distribution(self, triples):
+        """Calculate a slightly shifted target distribution by adding random noise."""
+
+        relation_triples = dsutils.group_triples(triples, group_type="relation")
+        original_distribution = {relation: 1.0 for relation, _ in relation_triples.items()}
+
```
Comment (Prateek-Ch) [inline kge/util/distribution-shift/custom_distribution.py:17]: Target Distribution for Testing:
The target distribution you used to test might be too far from the training distribution. After you reminded me that sampling a sufficient amount of triples isn’t always feasible, I realized we should avoid creating a test set with a large distribution shift. Instead, we can evaluate whether a very small distribution shift affects performance or increases variance first.
A more convenient way to define the target distribution would be as follows:

    Use the original i.i.d. distribution and add a small random noise to the vectors (e.g., randomly increase or decrease values by a small amount).
    This approach will also scale better for large distribution vectors (e.g., with lengths of 30k+), where manual input isn’t practical.

File: kge/util/distribution-shift/custom_distribution.py [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+        relation_triples = dsutils.group_triples(all_triples, group_type="relation")
+        target_distribution = self.calculate_target_distribution(all_triples)
+
+        # Calculate target number of triples per relation
+        target_counts = {relation: int(len(all_triples) * prob) for relation, prob in target_distribution.items()}
+
+        # Sample triples for each relation
+        train_valid_triples = []
+        for relation, count in target_counts.items():
+            available_triples = relation_triples[relation]
```
Comment (Prateek-Ch) [inline kge/util/distribution-shift/custom_distribution.py:41]: Ensuring Triple Availability for Relations:
In line 26 of custom_distribution.py, you ensure that the number of sampled triples doesn’t exceed the available triples. However, it’s also crucial to leave at least one triple per relation in the training set. This is necessary to create embeddings for each relation. Without this, we’d effectively switch to a different KGE setting (inductive link prediction), where there are unseen entities or relations in the test set.

File: kge/util/distribution-shift/custom_distribution.py [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+    checkpoint_path = "local/experiments/20241119-072058-wnrr-rescal/checkpoint_best.pt"
+    output_folder = "data/custom-dataset-updated"
+    noise_magnitude = 0.05
+
+    # Initialize distribution class
+    custom_distribution = CustomDistribution(checkpoint_path, noise_magnitude)
+
+    # Create training and validation splits
+    train_triples, valid_triples = custom_distribution.sample_training_validation_set()
+
```
Comment (Prateek-Ch) [inline kge/util/distribution-shift/custom_distribution.py:75]: Maintaining Training-Test Split Ratios:
Ensure that the ratio of training to test data remains consistent with the original split, where the test set is significantly smaller. This would leave more room for introducing larger distribution shifts while maintaining the integrity of the dataset.
